pkg/finality-grandpa/past_rounds.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/constraints"
+)
+
+// wraps a voting round with a new future that resolves when the round can
+// be discarded from the working set.
+//
+// that point is when the round-estimate is finalized.
+type backgroundRound[
+	Hash constraints.Ordered, Number constraints.Unsigned, Signature comparable,
+	ID constraints.Ordered, E Environment[Hash, Number, Signature, ID],
+] struct {
+	inner           votingRound[Hash, Number, Signature, ID, E]
+	finalizedNumber Number
+	roundCommitter  *roundCommitter[Hash, Number, Signature, ID, E]
+
+	waker *waker
+}
+
+func (br *backgroundRound[Hash, Number, Signature, ID, E]) roundNumber() uint64 {
+	return br.inner.roundNumber()
+}
+
+func (br *backgroundRound[Hash, Number, Signature, ID, E]) votingRound() votingRound[Hash, Number, Signature, ID, E] {
+	return br.inner
+}
+
+func (br *backgroundRound[Hash, Number, Signature, ID, E]) isDone() bool {
+	// no need to listen on a round anymore once the estimate is finalized.
+	//
+	// we map `br.roundCommitter == nil` to true because
+	//   - rounds are not backgrounded when incomplete unless we've skipped forward
+	//   - if we skipped forward we may never complete this round and we don't need
+	//     to keep it forever.
+	var ls = br.roundCommitter == nil
+	if !ls {
+		return false
+	}
+	var rs = true
+	estimate := br.inner.roundState().Estimate
+	if estimate != nil {
+		rs = estimate.Number <= br.finalizedNumber
+	}
+	return ls && rs
+}
+
+func (br *backgroundRound[Hash, Number, Signature, ID, E]) updateFinalized(newFinalized Number) {
+	switch {
+	case br.finalizedNumber >= newFinalized:
+	default:
+		br.finalizedNumber = newFinalized
+	}
+
+	// wake up the future to be polled if done.
+	if br.isDone() {
+		br.waker.wake()
+	}
+}
+
+type concluded uint64
+type committed[Hash, Number, Signature, ID any] Commit[Hash, Number, Signature, ID]
+
+type backgroundRoundChange[Hash, Number, Signature, ID any] struct {
+	variant any
```

Comment:
Should this be constrained to `concluded | committed`

Reply:
It is in the `setBackgroundRoundChangeVariant` function.

Reply:
I revised this a little introducing `newBackgroundRoundChange`.